client-runtime/src/main/java/com/microsoft/rest/v2/http/NettyClient.java

```diff
@@ -172,306 +187,665 @@ private NettyAdapter(int eventLoopGroupSize, int channelPoolSize) {
+                if (request.body() == null) {
+                    writeBodyEnd();
+                } else {
+                    request.body().subscribe(new RequestSubscriber(inboundHandler));
+                }
+            } catch (Exception e) {
+                emitError(e);
+            }
+        }
 
-                        if (request.httpMethod() == com.microsoft.rest.v2.http.HttpMethod.HEAD) {
-                            // Use HttpClientCodec for HEAD operations
-                            if (channel.pipeline().get("HttpClientCodec") == null) {
-                                channel.pipeline().remove(HttpRequestEncoder.class);
-                                channel.pipeline().replace(HttpResponseDecoder.class, "HttpClientCodec", new HttpClientCodec());
-                            }
-                        } else {
-                            // Use HttpResponseDecoder for other operations
-                            if (channel.pipeline().get("HttpResponseDecoder") == null) {
-                                channel.pipeline().replace(HttpClientCodec.class, "HttpResponseDecoder", new HttpResponseDecoder());
-                                channel.pipeline().addAfter("HttpResponseDecoder", "HttpRequestEncoder", new HttpRequestEncoder());
-                            }
-                        }
+        private final class RequestSubscriber implements FlowableSubscriber<ByteBuffer> {
+            Subscription subscription;
+
+            // we need a done flag because an onNext emission can throw and emit an Error
+            // event though the onNext cancels the subscription that is best endeavours for the
+            // upstream so we need to be defensive about terminal events that follow
+            private boolean done;
+
+            private final HttpClientInboundHandler inboundHandler;
+
+            RequestSubscriber(HttpClientInboundHandler inboundHandler) {
+                this.inboundHandler = inboundHandler;
+            }
 
-                        final DefaultHttpRequest raw = new DefaultHttpRequest(HttpVersion.HTTP_1_1,
-                                HttpMethod.valueOf(request.httpMethod().toString()),
-                                request.url().toString());
+            @Override
+            public void onSubscribe(Subscription s) {
+                subscription = s;
+                inboundHandler.requestContentSubscription = subscription;
+                subscription.request(1);
+            }
 
-                        for (HttpHeader header : request.headers()) {
-                            raw.headers().add(header.name(), header.value());
+            GenericFutureListener<Future<Void>> onChannelWriteComplete = (Future<Void> future) -> {
+                if (!future.isSuccess()) {
+                    done = true;
+                    subscription.cancel();
+                    emitError(future.cause());
+                }
+                else if (channel.isWritable()) {
```

Comment:
Trouble is this could stall the read of the request body. If we put this in then we need to detect the channel write status change from true to false and then request if the latest write has completed. A bit of coordination between onChannelWriteComplete and writability state change that might need a drain approach.

Reply:
I noticed that we do request when channel writability changes but can over-request there too. I'll sort this out. Just need a CAS loop in AcquisitionListener.